src/ucp/tag/offload.c

```diff
@@ -78,62 +89,98 @@ void ucp_tag_offload_completed(uct_tag_context_t *self, uct_tag_t stag,
     ucp_request_complete_tag_recv(req, status);
 }
 
-static size_t ucp_tag_offload_fetch_rkey(ucp_ep_t *ep, ucp_sw_rndv_hdr_t *sw_hdr,
-                                         ucp_rndv_rts_hdr_t *rts)
+static UCS_F_ALWAYS_INLINE size_t
+ucp_tag_offload_rkey_size(ucp_context_t *ctx)
 {
-    uint64_t *addr;
-    size_t rkey_size;
+    ucp_worker_iface_t *iface = ucp_tag_offload_iface(ctx);
+    ucp_md_index_t md_idx;
+
+    md_idx = ctx->tl_rscs[iface->rsc_index].md_index;
+    return ctx->tl_mds[md_idx].attr.rkey_packed_size;
+}
 
-    if (!(sw_hdr->flags & UCP_RNDV_RTS_FLAG_PACKED_RKEY) || ucp_ep_is_stub(ep)) {
-        rts->address = 0;
+static UCS_F_ALWAYS_INLINE size_t
+ucp_tag_offload_copy_rkey(ucp_rndv_rts_hdr_t *rts, const void *rkey_buf,
+                          size_t rkey_size)
+{
+    if (rkey_buf == NULL) {
         return 0;
     }
 
-    /* SW RNDV with valid remote addr and key. Must be large message which
-     * does not fit offload RNDV proto. */
-    rkey_size    = ucp_ep_md_attr(ep, ucp_ep_get_tag_lane(ep))->rkey_packed_size;
-    addr         = (uint64_t*)(sw_hdr + 1);
-    rts->address = *addr;
-    rts->flags  |= sw_hdr->flags | UCP_RNDV_RTS_FLAG_OFFLOAD;
-    memcpy(rts + 1, addr + 1, rkey_size);
+    ucs_assert(rts != NULL);
+    memcpy(rts + 1, rkey_buf, rkey_size);
+    rts->flags |= UCP_RNDV_RTS_FLAG_OFFLOAD | UCP_RNDV_RTS_FLAG_PACKED_RKEY;
     return rkey_size;
 }
 
-/* RNDV request matched by the transport. Need to proceed with AM based RNDV */
+static UCS_F_ALWAYS_INLINE void
+ucp_tag_offload_fill_rts(ucp_rndv_rts_hdr_t *rts, const ucp_request_hdr_t *hdr,
+                         uint64_t stag, uint64_t addr, size_t size,
+                         uint16_t flags)
+{
+    rts->super.tag = stag;
+    rts->sreq      = *hdr;
+    rts->address   = addr;
+    rts->size      = size;
+    rts->flags     = flags;
+}
+
+static UCS_F_ALWAYS_INLINE size_t
+ucp_tag_offload_fill_sw_rts(ucp_rndv_rts_hdr_t *rts,
+                            const ucp_sw_rndv_hdr_t *sw_rndv_hdr,
+                            unsigned header_length, uint64_t stag,
+                            size_t rkey_size)
+{
+    size_t length = sizeof(*rts);
+    uint64_t *addr_ptr;
+
+    if (sw_rndv_hdr->flags & UCP_RNDV_RTS_FLAG_PACKED_RKEY) {
+        ucs_assert(rkey_size);
+        ucs_assert((sizeof(*sw_rndv_hdr) + rkey_size + sizeof(uint64_t)) ==
+                   header_length);
+        addr_ptr = (uint64_t*)(sw_rndv_hdr + 1);
```

Comment:
maybe define another struct to contain ucp_sw_rndv_hdr_t+address, for sw rndv ? seems better than pointer arithmetic..